hax: IndexError: list index out of range when Preparing the reply for HaNvecGetEvent

On SAGE cluster when starting any client process (like c0cp or mcp) it gets stuck forever with the following stack:

```
#3  0x00007fab6d941014 in m0_semaphore_down (semaphore=0x7ffd9ffe3158) at lib/user_space/semaphore.c:62
#4  0x00007fab6d933b15 in m0_chan_wait (link=link@entry=0x7ffd9ffe3130) at lib/chan.c:341
#5  0x00007fab6d8c2994 in m0_conf_objs_ha_update (nvec=nvec@entry=0x7ffd9ffe3270) at conf/helpers.c:576
#6  0x00007fab6d8c2cc8 in m0_conf_confc_ha_update_async (confc=<optimized out>, nvec=nvec@entry=0x7ffd9ffe3270, chan=chan@entry=0x0) at conf/helpers.c:642
#7  0x00007fab6d8c2d62 in m0_conf_confc_ha_update (confc=<optimized out>) at conf/helpers.c:663
#8  0x00007fab6d95baba in confc_init (m0c=0x2107d00) at motr/client_init.c:939
#9  initlift_confc (mach=0x2111138) at motr/client_init.c:1008
#10 0x00007fab6d9ce41c in state_set (mach=0x2111138, state=5, rc=<optimized out>) at sm/sm.c:457
#11 0x00007fab6d95c98c in initlift_move_next_floor (m0c=m0c@entry=0x2107d00) at motr/client_init.c:364
#12 0x00007fab6d95d0c0 in m0_client_init (m0c_p=0x801468 <instance>, conf=0x801420 <conf>, init_m0=<optimized out>) at motr/client_init.c:1568
```

In the system log there is following error from hax:

```
Jan 11 16:04:26 client-25 hare-hax[16976]: 2021-01-11 16:04:26,545 [DEBUG] {qconsumer} Preparing the reply for HaNvecGetEvent (nvec size = 976)
Jan 11 16:04:26 client-25 hare-hax[16976]: 2021-01-11 16:04:26,596 [ERROR] {qconsumer} **ERROR**
Jan 11 16:04:26 client-25 hare-hax[16976]: Traceback (most recent call last):
Jan 11 16:04:26 client-25 hare-hax[16976]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/motr/__init__.py", line 42, in wrapper
Jan 11 16:04:26 client-25 hare-hax[16976]: return fn(*args, **kwargs)
Jan 11 16:04:26 client-25 hare-hax[16976]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/motr/__init__.py", line 278, in ha_nvec_get_reply
Jan 11 16:04:26 client-25 hare-hax[16976]: n.note.no_id.f_key) !=
Jan 11 16:04:26 client-25 hare-hax[16976]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 414, in get_conf_obj_status
Jan 11 16:04:26 client-25 hare-hax[16976]: return self.get_node_health(node_name)
Jan 11 16:04:26 client-25 hare-hax[16976]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 435, in get_node_health
Jan 11 16:04:26 client-25 hare-hax[16976]: return str(node_data[0]['Status'])
Jan 11 16:04:26 client-25 hare-hax[16976]: IndexError: list index out of range
```

(Full log - [hax-IndexError-on-HaNvecGetEvent-syslog.txt](https://github.com/Seagate/cortx-hare/files/5797457/hax-IndexError-on-HaNvecGetEvent-syslog.txt).)


It was working fine yesterday, but not today. Maybe it has something to do with the fact that a few client nodes were shutdown:

```
$ consul members
Node          Address           Status  Type    Build  Protocol  DC   Segment
sage-tier1-1  172.18.1.1:8301   alive   server  1.7.8  2         dc1  <all>
client-21     172.18.1.21:8301  alive   client  1.7.8  2         dc1  <default>
client-22     172.18.1.22:8301  left    client  1.7.8  2         dc1  <default>
client-23     172.18.1.23:8301  alive   client  1.7.8  2         dc1  <default>
client-24     172.18.1.24:8301  alive   client  1.7.8  2         dc1  <default>
client-25     172.18.1.25:8301  alive   client  1.7.8  2         dc1  <default>
client-26     172.18.1.26:8301  alive   client  1.7.8  2         dc1  <default>
client-27     172.18.1.27:8301  alive   client  1.7.8  2         dc1  <default>
client-28     172.18.1.28:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-01   172.18.1.33:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-02   172.18.1.34:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-03   172.18.1.35:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-04   172.18.1.36:8301  left    client  1.7.8  2         dc1  <default>
sage-tier1-2  172.18.1.2:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-1  172.18.1.3:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-2  172.18.1.4:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-4  172.18.1.6:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier3-1  172.18.1.7:8301   alive   client  1.7.8  2         dc1  <default>
visnode-01    172.18.1.37:8301  alive   client  1.7.8  2         dc1  <default>
visnode-02    172.18.1.38:8301  alive   client  1.7.8  2         dc1  <default>
visnode-03    172.18.1.39:8301  alive   client  1.7.8  2         dc1  <default>
visnode-04    172.18.1.40:8301  left    client  1.7.8  2         dc1  <default>
```

But such things should not affect the cluster or the other client nodes which are alive.

Versions:

Hare: cortx-hare-1.0.0-1_git28f3372.el7.x86_64
Motr: cortx-motr-1.0.0-1_git89f7737_3.10.0_1127.19.1.el7.x86_64 (from https://github.com/Seagate/cortx-motr/pull/304)

@knekrasov, I am trying to build this patch,
```
diff --git a/hax/hax/util.py b/hax/hax/util.py
index 4473917..382fe54 100644
--- a/hax/hax/util.py
+++ b/hax/hax/util.py
@@ -446,9 +446,23 @@ class ConsulUtil:
             if f'{fid}' == x['Key'].split('/')[-1]
         ]

+    def is_node_alive(self, node: str) -> bool:
+        try:
+            members_data = self.cns.agent.members()
+            LOG.debug("members: %s", members_data)
+            for member in members_data:
+                if member['Name'] == node:
+                    return int(member['Status']) == 1
+            return True
+        except (ConsulException, HTTPError, RequestException) as e:
+            raise HAConsistencyException(
+                'Failed to members data from Consul') from e
+
     def get_node_health(self, node: str) -> str:
         try:
             node_data = self.cns.health.node(node)[1]
+            if not node_data and (not self.is_node_alive(node)):
+                return 'failed'
             return str(node_data[0]['Status'])
         except (ConsulException, HTTPError, RequestException) as e:
             raise HAConsistencyException(
```
but failing with below build error,
```
WARNING: flake8 setuptools integration is deprecated and scheduled for removal in 4.x.  For more information, see https://gitlab.com/pycqa/flake8/issues/544
running mypy
hax/util.py:451: error: "Agent" has no attribute "members"
Found 1 error in 1 file (checked 18 source files)

error: Mypy returned 1. Exiting with FAILURE status
make[1]: *** [check-hax] Error 1
make: *** [build] Error 2
```
but I can see the function part of Agent class in the python-consul sources also able to invoke it explicitly in python3 interactive shell.
Tried multiple things but somehow none of the Agent routines are available other than `self()`, I suspect some kind of build problem but not able to figure it out yet. Kindly check the same. Thank you.

Now all the nodes are alive, but the issue is still there:

```
$ consul members
Node          Address           Status  Type    Build  Protocol  DC   Segment
sage-tier1-1  172.18.1.1:8301   alive   server  1.7.8  2         dc1  <all>
client-21     172.18.1.21:8301  alive   client  1.7.8  2         dc1  <default>
client-22     172.18.1.22:8301  alive   client  1.7.8  2         dc1  <default>
client-23     172.18.1.23:8301  alive   client  1.7.8  2         dc1  <default>
client-24     172.18.1.24:8301  alive   client  1.7.8  2         dc1  <default>
client-25     172.18.1.25:8301  alive   client  1.7.8  2         dc1  <default>
client-26     172.18.1.26:8301  alive   client  1.7.8  2         dc1  <default>
client-27     172.18.1.27:8301  alive   client  1.7.8  2         dc1  <default>
client-28     172.18.1.28:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-01   172.18.1.33:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-02   172.18.1.34:8301  alive   client  1.7.8  2         dc1  <default>
datawarp-03   172.18.1.35:8301  alive   client  1.7.8  2         dc1  <default>
sage-tier1-2  172.18.1.2:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-1  172.18.1.3:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-2  172.18.1.4:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier2-4  172.18.1.6:8301   alive   client  1.7.8  2         dc1  <default>
sage-tier3-1  172.18.1.7:8301   alive   client  1.7.8  2         dc1  <default>
visnode-01    172.18.1.37:8301  alive   client  1.7.8  2         dc1  <default>
visnode-02    172.18.1.38:8301  alive   client  1.7.8  2         dc1  <default>
visnode-03    172.18.1.39:8301  alive   client  1.7.8  2         dc1  <default>
```

The same error from hax when running `mcp`:

```
Jan 12 10:33:41 client-22 hare-hax[20538]: 2021-01-12 10:33:41,887 [DEBUG] {qconsumer} Preparing the reply for HaNvecGetEvent (nvec size = 976)
Jan 12 10:33:41 client-22 hare-hax[20538]: 2021-01-12 10:33:41,888 [DEBUG] {qconsumer} Attempting to invoke the repeatable call: get_conf_obj_status
Jan 12 10:33:41 client-22 hare-hax[20538]: 2021-01-12 10:33:41,936 [ERROR] {qconsumer} **ERROR**
Jan 12 10:33:41 client-22 hare-hax[20538]: Traceback (most recent call last):
Jan 12 10:33:41 client-22 hare-hax[20538]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/motr/__init__.py", line 40, in wrapper
Jan 12 10:33:41 client-22 hare-hax[20538]: return fn(*args, **kwargs)
Jan 12 10:33:41 client-22 hare-hax[20538]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/motr/__init__.py", line 230, in ha_nvec_get_reply
Jan 12 10:33:41 client-22 hare-hax[20538]: n.note.no_id.f_key) !=
Jan 12 10:33:41 client-22 hare-hax[20538]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 117, in wrapper
Jan 12 10:33:41 client-22 hare-hax[20538]: result = f(*args, **kwds)
Jan 12 10:33:41 client-22 hare-hax[20538]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 389, in get_conf_obj_status
Jan 12 10:33:41 client-22 hare-hax[20538]: return self.get_node_health(node_name)
Jan 12 10:33:41 client-22 hare-hax[20538]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 410, in get_node_health
Jan 12 10:33:41 client-22 hare-hax[20538]: return str(node_data[0]['Status'])
Jan 12 10:33:41 client-22 hare-hax[20538]: IndexError: list index out of range
```

--- a/hax/hax/util.py
+++ b/hax/hax/util.py
@@ -489,10 +489,53 @@ class ConsulUtil:
             if f'{fid}' == x['Key'].split('/')[-1]
         ]
 
+    def is_node_alive(self, node: str) -> bool:
+        """
+        Checks via Consul Members API whether the given node is alive.
+        """
+        try:
+            # Returns data of the following kind:
+            # [{
+            #     'Name': 'localhost',
+            #     'Addr': '192.168.6.214',
+            #     'Port': 8301,
+            #     'Tags': {
+            #         'acls': '0',
+            #         'bootstrap': '1',
+            #         'build': '1.7.8:9a5a1218',
+            #         'dc': 'dc1',
+            #         'id': 'dd8a91f6-ca32-30e0-983c-8f309d653045',
+            #         'port': '8300',
+            #         'raft_vsn': '3',
+            #         'role': 'consul',
+            #         'segment': '',
+            #         'vsn': '2',
+            #         'vsn_max': '3',
+            #         'vsn_min': '2',
+            #         'wan_join_port': '8302'
+            #     },
+            #     'Status': 1,
+            #     'ProtocolMin': 1,
+            #     'ProtocolMax': 5,
+            #     'ProtocolCur': 2,
+            #     'DelegateMin': 2,
+            #     'DelegateMax': 5,
+            #     'DelegateCur': 4
+            # }]
+            members_data = self.cns.agent.members()
+            LOG.log(TRACE, "members: %s", members_data)
+            for member in members_data:
+                if member['Name'] == node:
+                    return int(member['Status']) == 1
+            return True
+        except (ConsulException, HTTPError, RequestException) as e:
+            raise HAConsistencyException(
+                'Failed to members data from Consul') from e
+
     def get_node_health(self, node: str) -> str:
         try:
             node_data = self.cns.health.node(node)[1]
-            if not node_data:
+            if not node_data or (not self.is_node_alive(node)):
                 return 'failed'
             return str(node_data[0]['Status'])
         except (ConsulException, HTTPError, RequestException) as e:

--- a/stubs/consul/__init__.pyi
+++ b/stubs/consul/__init__.pyi
@@ -76,6 +76,7 @@ class Catalog:
 
 class Agent:
     def self(self) -> Dict[str, Any]: ...
+    def members(self, wan: bool = False) -> List[Dict[str, Any]]: ...
 
 class KV:
     def get(

